### PR TITLE
NEXUS-8914: Proxy fetches more than should

### PR DESCRIPTION
**Pull request by cstamas**
This PR solves the "proxy fetches more than should" issue all 3 problems by fixing formats with regard of handling properties like lastVerified, lastModified and ETag, needed for proper and sane upstream proxy functionality. Interestingly, this implied immediate improvements for downstream content serving too (ie. support caching UAs), as it all revolves around Content and it's attributes.

Changes:
* Fix for NEXUS-8914 1. case (detected by PR #1481): more remote fetches happens than needed, CacheInfo is saved not only for 304 response.
* Fix for NEXUS-8914 2. case: save proxy provided content attributes (lastVerified, lastModified, etag) immediately, along with asset and content itself. No need for separate call.
* Fix for NEXUS-8914 3. case: Make lastModified/etag handling generic, format independent, so Raw does not refetch same content over and over again (from upstream). For this to achieve, drop RAW format specific RawContent, make it use Content instead (which is almost equivalent of RawContent)
* Implication of that above: NEXUS-8923: Make RAW repo emit proper headers and support conditional GETs (before all mvn sites were re-downloaded every time accessed), as now with Content used, it's only about adding ContentHeadersHandler and ConditionalRequestHandlers to it
* Similar changes applies to NuGet, thus it made also emit std HTTP headers and capable of conditional GETs. ie. nuget proxy will now pass along downstream the ETag that upstream gallery sends, same for lastModified (changes were needed to be propagated due to ProxyFacetSupport changes)

On top of that all above:
* NEXUS-8817: Due to improvements above (basically addition of CacheInfo), implementing proxy cache invalidation became trivial: it is generic, and all formats have it now exposed via ProxyFacet

Issue
https://issues.sonatype.org/browse/NEXUS-8914

CI
http://bamboo.s/browse/NX3-OSSF491

### CODE REVIEW

File: components/nexus-common/src/main/java/org/sonatype/nexus/common/time/DateHelper.java [hunk 1]
```diff
@@ -1,6 +1,6 @@
 /*
  * Sonatype Nexus (TM) Open Source Version
- * Copyright (c) 2008-present Sonatype, Inc.
+ * Copyright (c) 2008-2015 Sonatype, Inc.
```
Comment (mcculls) [inline components/nexus-common/src/main/java/org/sonatype/nexus/common/time/DateHelper.java]: Remember to rebase/merge with latest master and re-generate the headers with `mvn license:format -N`

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/CacheInfo.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+/**
+ * A proxy cache auxiliary info for given element. {@link ProxyFacetSupport} relies on information provided by this
+ * class to implement "aging" and proxy cache invalidation.
+ *
+ * @since 3.0
+ */
+public class CacheInfo
+{
+  private static final String P_PROXY_CACHE = "proxy_cache";
```
Comment (mrprescott) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/CacheInfo.java]: Maybe a sprinkle of javadoc to indicate that P_PROXY_CACHE is a map, while the others are properties? CACHE_TOKEN in particular could mean anything, but I think it's always an Etag?  ('remote cache token'?)

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/ProxyFacetSupport.java [hunk 3]
```diff
@@ -86,6 +86,8 @@ public String toString() {
 
   private boolean remoteUrlChanged;
 
+  private String cacheToken;
```
Comment (mrprescott) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/ProxyFacetSupport.java]: Consider concurrency control around this - volatile or AtomicReference or whatever.

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/CacheInfo.java (hunk 2 above)
Comment (cstamas) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/CacheInfo.java]: Ack.

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/ProxyFacetSupport.java (hunk 3 above)
Comment (mrprescott) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/ProxyFacetSupport.java]: As discussed in the call this needs to be persisted, but perhaps that's another PR.

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/ContentMarshaller.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  private static final String P_CONTENT = "content";
+
+  private static final String P_LAST_MODIFIED = "last_modified";
+
+  private static final String P_ETAG = "etag";
+
+  private ContentMarshaller() {}
+
+  /**
+   * Extracts non-format specific content attributes into the passed in {@link AttributesMap} (usually originating from
```
Comment (mrprescott) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/ContentMarshaller.java]: Javadoc out of date; you're not passing in AttributesMap any more.

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/ContentMarshaller.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+   * {@link Content#getAttributes()}) from passed in {@link Asset} and format required hashes.
+   */
+  public static void extract(final AttributesMap content,
```
Comment (cstamas) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/ContentMarshaller.java]: Rename method, name it more explicitly, and rename params too (at some point it was `Content` being passed in here)

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/ContentMarshaller.java [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    content.set(Content.CONTENT_ETAG, etag);
+    content.set(Content.CONTENT_HASH_CODES_MAP, hashCodes);
+    content.set(CacheInfo.class, CacheInfo.extract(asset));
+  }
+
+  /**
+   * Applies non-format specific content attributes onto passed in {@link Asset} from passed in {@link AttributesMap}
+   * (usually originating from {@link Content#getAttributes()}).
+   */
+  public static void apply(final Asset asset, final AttributesMap content) {
```
Comment (cstamas) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/ContentMarshaller.java]: Rename method, name it more explicitly, and rename params too (at some point it was `Content` being passed in here)

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/CacheInfo.java [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+        ", cacheToken='" + cacheToken + '\'' +
+        '}';
+  }
+
+  /**
+   * Helper method that extracts {@link CacheInfo} from an {@link Asset}. Input cannot be {@code null}, but returned
+   * value might be, if passed in {@link Asset} does not contain cache-info properties.
+   */
+  @Nullable
+  public static CacheInfo extract(final Asset asset) {
```
Comment (cstamas) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/CacheInfo.java]: Rename method, name it more explicitly.

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/CacheInfo.java [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+      return null;
+    }
+    final String cacheToken = proxyCache.get(P_CACHE_TOKEN, String.class);
+    return new CacheInfo(lastVerified, cacheToken);
+  }
+
+  /**
+   * Helper method that applies {@link CacheInfo} to an {@link Asset}. None of the inputs may be {@code null}.
+   */
+  public static void apply(final Asset asset, final CacheInfo cacheInfo) {
```
Comment (cstamas) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/CacheInfo.java]: Rename method, name it more explicitly.

File: plugins/basic/nexus-repository-nuget/src/main/java/com/sonatype/nexus/repository/nuget/internal/NugetGalleryFacetImpl.java [hunk 9]
```diff
@@ -607,13 +578,22 @@ void maintainAggregateInfo(final StorageTx storageTx, final Iterable<Component>
...
-                                              final Map<String, String> data)
+                                              final Map<String, String> data,
+                                              @Nullable final AttributesMap contentAttributes)
   {
     try {
       Asset asset = findOrCreateAsset(storageTx, component);
       updateAssetMetadata(asset, data, component.isNew());
 
-      asset.formatAttributes().set(P_LAST_VERIFIED_DATE, new Date());
+      if (contentAttributes != null) {
```
Comment (mrprescott) [inline plugins/basic/nexus-repository-nuget/src/main/java/com/sonatype/nexus/repository/nuget/internal/NugetGalleryFacetImpl.java]: These same 8 lines appear in the maven codebase, too. Refactor?

**Comment by mrprescott** [comment]
+1 with minor comments

File: plugins/basic/nexus-repository-raw/src/main/java/org/sonatype/nexus/repository/raw/internal/RawContentFacetImpl.java [hunk 10]
```diff
@@ -110,8 +111,17 @@ public Void execute(final StorageTx tx) {
               asset = tx.firstAsset(component);
             }
 
-            asset.formatAttributes().set(P_LAST_VERIFIED_DATE, new Date());
-            tx.setBlob(asset, path, streamSupplier.get(), hashAlgorithms, null, content.getContentType());
+            if (payload instanceof Content) {
```
Comment (mrprescott) [inline plugins/basic/nexus-repository-raw/src/main/java/org/sonatype/nexus/repository/raw/internal/RawContentFacetImpl.java]: I think this, too, is basically a logical duplicate, just masked by the slightly different formulation. But the basic idea is the same. Actually, does this version accidentally omit setting LAST_MODIFIED if the content attributes don't include it?  The others did that.

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/ProxyFacetSupport.java (hunk 3 above)
Comment (cstamas) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/ProxyFacetSupport.java]: Will not persist this in this PR, should be easy to do in another PR.

File: plugins/basic/nexus-repository-raw/src/main/java/org/sonatype/nexus/repository/raw/internal/RawContentFacetImpl.java (hunk 10 above)
Comment (cstamas) [inline plugins/basic/nexus-repository-raw/src/main/java/org/sonatype/nexus/repository/raw/internal/RawContentFacetImpl.java]: Yup, this is wrong here.

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/CacheInfo.java [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+package org.sonatype.nexus.repository.proxy;
+
+import javax.annotation.Nullable;
+
+import org.joda.time.DateTime;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * A proxy cache auxiliary info for given element. {@link ProxyFacetSupport} relies on information provided by this
```
Comment (mcculls) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/CacheInfo.java]: Element isn't mentioned in the class, perhaps "Maintains cache details for proxied resources..."?

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/CacheInfoMarshaller.java [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+import static com.google.common.base.Preconditions.checkNotNull;
+import static org.sonatype.nexus.common.time.DateHelper.toDate;
+import static org.sonatype.nexus.common.time.DateHelper.toDateTime;
+
+/**
+ * Extracts non-format specific cache related metadata from asset.
+ *
+ * @since 3.0
+ */
+public class CacheInfoMarshaller
```
Comment (mcculls) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/CacheInfoMarshaller.java]: This class just contains static methods/fields that could be moved into `CacheInfo`

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/ProxyFacetSupport.java (hunk 3 above)
Comment (mcculls) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/ProxyFacetSupport.java]: Was @mrprescott's question about concurrency resolved? (since persistable != thread-safe)

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/ProxyFacetSupport.java [hunk 13]
```diff
@@ -262,36 +270,41 @@ private String extractETag(final Header etagHeader) {
...
-  private boolean isStale(final Context context) throws IOException {
+  private boolean isStale(final Context context, final Content content) throws IOException {
+    if (content == null) {
+      // not in cache
+      return true;
+    }
+
     if (config.artifactMaxAge < 0) {
-      log.trace("Artifact max age checking disabled");
+      log.trace("Content max age checking disabled");
```
Comment (mcculls) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/ProxyFacetSupport.java]: Should the property be changed to match the message?

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/ProxyFacetSupport.java [hunk 14]
```diff
@@ -262,36 +270,41 @@ private String extractETag(final Header etagHeader) {
...
       return false;
     }
 
-    final DateTime lastUpdated = getCachedPayloadLastVerified(context);
-
-    if (lastUpdated == null) {
-      log.debug("Artifact last modified date unknown");
+    // FIXME: maybe require(CacheInfo.class), as cache content w/o CacheInfo should not exists in the first place
```
Comment (mcculls) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/ProxyFacetSupport.java]: Any reason not do this now vs. postponing it?

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/ProxyFacetSupport.java [hunk 15]
```diff
@@ -262,36 +270,41 @@ private String extractETag(final Header etagHeader) {
...
+    // FIXME: what happens with cache content once we go distributed?
```
Comment (mcculls) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/ProxyFacetSupport.java]: There are various places which will need reviewing when we do that, so not sure we need to leave this comment here

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/ProxyFacetSupport.java [hunk 16]
```diff
@@ -262,36 +270,41 @@ private String extractETag(final Header etagHeader) {
...
+    final CacheInfo cacheInfo = content.getAttributes().get(CacheInfo.class);;
+    if (cacheInfo == null) {
+      log.debug("Content cache info unknown");
       return true;
     }
-
-    final DateTime earliestFreshDate = new DateTime().minusMinutes(config.artifactMaxAge);
-    return lastUpdated.isBefore(earliestFreshDate);
+    if (cacheToken != null && !cacheToken.equals(cacheInfo.getCacheToken())) {
+      log.debug("Content expired (cacheToken)");
```
Comment (mcculls) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/ProxyFacetSupport.java]: Log the expired cache details along with the content path

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/ProxyFacetSupport.java [hunk 17]
```diff
@@ -262,36 +270,41 @@ private String extractETag(final Header etagHeader) {
...
+      return true;
+    }
+    if (cacheInfo.getLastVerified().isBefore(new DateTime().minusMinutes(config.artifactMaxAge))) {
+      log.debug("Content expired (age)");
```
Comment (mcculls) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/ProxyFacetSupport.java]: As above, fixed message without context isn't that useful

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/ContentMarshaller.java [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+import static com.google.common.base.Preconditions.checkNotNull;
+import static org.sonatype.nexus.common.time.DateHelper.toDate;
+import static org.sonatype.nexus.common.time.DateHelper.toDateTime;
+
+/**
+ * Extracts non-format specific auxiliary metadata from asset.
+ *
+ * @since 3.0
+ */
+public class ContentMarshaller
```
Comment (mcculls) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/ContentMarshaller.java]: This name suggests that it marshals content, but it actually marshals auxiliary metadata.

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/ContentMarshaller.java [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+{
+  /**
+   * Key of the nested map on {@link Asset} attributes carrying content related attributes.
+   */
+  private static final String P_CONTENT = "content";
```
Comment (mcculls) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/ContentMarshaller.java]: As above, this name doesn't match what it actually stores

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/ContentMarshaller.java [hunk 20]
```diff
@@ -0,0 +1,123 @@
...
+  /**
+   * Applies non-format specific content attributes onto passed in {@link Asset} from passed in {@link AttributesMap}
+   * (usually originating from {@link Content#getAttributes()}). The {@code contentAttributes} input may be null too,
+   * in which case this method will create empty content attributes. If the content attributes (passed in or new) does
+   * not contains {@link Content#CONTENT_LAST_MODIFIED} mapping, mapping with "now" timestamp will be added.
+   *
+   * @see #applyToAsset(Asset, AttributesMap)
+   */
+  public static void applyToAssetMaintainLastModified(final Asset asset,
+                                                      @Nullable final AttributesMap contentAttributes)
```
Comment (mcculls) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/ContentMarshaller.java]: Method name suggests this is trying to do too much - how about a method `maintainLastModified` which takes an asset and attributes map and just maintains the last-modified time, which clients can then apply using `applyToAsset`.

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/ProxyFacetSupport.java (hunk 3 above)
Comment (cstamas) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/ProxyFacetSupport.java]: @mcculls No, nothing yet. The "concurrency" issue here is that ongoing content request might still be treated as non-expired, but unsure how "strict" we want to be regarding expiring caches. IMHO, this is a non-issue, but we can add "volatile" at least if we want to.

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/ProxyFacetSupport.java (hunk 16 above)
Comment (cstamas) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/ProxyFacetSupport.java]: CacheDetails OK, but not all reposes have "path"

Comment (mcculls) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/ProxyFacetSupport.java]: OK, anything else we can log to help tie the expired cache to the resource? Just thinking it might make it easier to trace things when there are interleaved requests.

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/ContentMarshaller.java (hunk 20 above)
Comment (cstamas) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/ContentMarshaller.java]: Ack

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/ProxyFacetSupport.java (hunk 13 above)
Comment (cstamas) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/ProxyFacetSupport.java]: Ack

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/ProxyFacetSupport.java (hunk 14 above)
Comment (cstamas) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/ProxyFacetSupport.java]: Ack

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/ProxyFacetSupport.java (hunk 15 above)
Comment (cstamas) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/ProxyFacetSupport.java]: Ack, removing and adding `require(CacheInfo.class)`

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/ProxyFacetSupport.java (hunk 17 above)
Comment (cstamas) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/ProxyFacetSupport.java]: On DEBUG level, along with thread-name, you could "follow" the fate of single request. As I said, am unsure what to add here, as not all assets have path/key/whatever, but am pretty sure that at DEBUG level you could see some line above this line (happening on same thread) that logs entering HTTP request fully (also at DEBUG level).

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/ContentMarshaller.java (hunk 18 above)
Comment (cstamas) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/ContentMarshaller.java]: This is same situation as CachaInfo + CacheInfoMarshaller. Are you suggesting to merge CacheInfoMarshaller into CacheInfo and ContentMarshaller into Content?

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/ContentMarshaller.java [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+{
+  /**
+   * Key of the nested map on {@link Asset} attributes carrying content related attributes.
+   */
+  private static final String P_CONTENT = "content";
+
+  /**
+   * Key of last modified {@link Date}.
+   */
+  private static final String P_LAST_MODIFIED = "last_modified";
```
Comment (kellyrob99) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/ContentMarshaller.java]: would prefer we try to keep some consistency here - above in Content we also have lastModified and etag constants indicating we're creating multiple locations/names for the same type of information. If the concepts are shared, IMO they should be extracted and used consistently by all creators/consumers rather than these subtle variations and duplications

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/ContentMarshaller.java (hunk 18 above)
Comment (kellyrob99) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/ContentMarshaller.java]: Agreed with @mcculls on the naming here, definitely misleading

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/ContentMarshaller.java (hunk 21 above)
Comment (cstamas) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/ContentMarshaller.java]: This key is how the value is persisted into OrientDB (basically Asset.content.last_modified and Asset.content.etag), and is the reason why this constant is private. I think it is good to keep them separate (this, orient record property key and Content attribute keys for future, like upgrades?)

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/ProxyFacetSupport.java (hunk 17 above)
Comment (mcculls) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/ProxyFacetSupport.java]: OK, as long as you can relate the different messages that should be enough

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/ContentMarshaller.java (hunk 18 above)
Comment (mcculls) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/ContentMarshaller.java]: @cstamas don't mind if you merge them or not (just pointing out that might be convenient to do so rather than have related methods sitting in a separate class) but the name definitely needs sorting out here as this doesn't marshal content, it only marshals content metadata.

File: plugins/basic/nexus-repository-nuget/src/main/java/com/sonatype/nexus/repository/nuget/internal/NugetGalleryFacetImpl.java [hunk 22]
```diff
@@ -244,7 +246,7 @@ protected void doPutContent(final String id, final String version, final Supplie
         component != null && tx.browseAssets(component).iterator().hasNext(),
         "Component metadata does not exist yet"
     );
-    createOrUpdateAssetAndContents(tx, component, streamSupplier.get(), null);
+    createOrUpdateAssetAndContents(tx, component, streamSupplier.get(), null, content.getAttributes());
```
Comment (kellyrob99) [inline plugins/basic/nexus-repository-nuget/src/main/java/com/sonatype/nexus/repository/nuget/internal/NugetGalleryFacetImpl.java]: if all we need here are the attributes, maybe change signature to take in just that?

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/ContentMarshaller.java (hunk 21 above)
Comment (kellyrob99) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/ContentMarshaller.java]: I can buy that, for now ;)

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/CacheInfo.java [hunk 23]
```diff
@@ -0,0 +1,61 @@
...
+ * class to implement "aging" and proxy cache invalidation.
+ *
+ * @since 3.0
+ */
+public class CacheInfo
+{
+  private final DateTime lastVerified;
+
+  @Nullable
+  private final String cacheToken;
```
Comment (kellyrob99) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/CacheInfo.java]: Under what condition would allowing null for this be sensible? Seemed that it was always set but I may have missed some scenario. In any case, it doesn't seem of value to allow it to be null

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/CacheInfoMarshaller.java [hunk 24]
```diff
@@ -0,0 +1,77 @@
...
+   */
+  private static final String P_CACHE_TOKEN = "cache_token";
+
+  private CacheInfoMarshaller() {}
+
+  /**
+   * Extracts non-format specific cache info from passed in {@link Asset}.
+   */
+  @Nullable
+  public static CacheInfo extractFromAsset(final Asset asset) {
```
Comment (kellyrob99) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/CacheInfoMarshaller.java]: needs UTs please

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/ContentMarshaller.java [hunk 25]
```diff
@@ -0,0 +1,121 @@
...
+   */
+  private static final String P_ETAG = "etag";
+
+  private ContentMarshaller() {}
+
+  /**
+   * Extracts non-format specific content attributes into the passed in {@link AttributesMap} (usually originating from
+   * {@link Content#getAttributes()}) from passed in {@link Asset} and format required hashes.
+   */
+  public static void extractFromAsset(final Asset asset,
```
Comment (kellyrob99) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/ContentMarshaller.java]: needs UTs please

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/ContentMarshaller.java (hunk 18 above)
Comment (cstamas) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/ContentMarshaller.java]: Ack, will kill CacheInfoMarshaller and ContentMarshaller and move those static into CacheInfo and Content classes.

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/CacheInfo.java (hunk 23 above)
Comment (cstamas) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/CacheInfo.java]: After NX3 boot this would be null. For persisting it, see discussion here https://github.com/sonatype/nexus-oss/pull/1502#discussion-diff-33868997

Am for leaving it as is for now, and in "persist cacheToken PR" we can then init it from configuration once persisted (or init it on 1st boot)".

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/CacheInfoMarshaller.java (hunk 24 above)
Comment (cstamas) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/CacheInfoMarshaller.java]: Ack, will do.

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/ContentMarshaller.java (hunk 25 above)
Comment (cstamas) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/ContentMarshaller.java]: Ack, will do.

**Comment by cstamas** [comment]
http://bamboo.s/browse/NX3-OSSF491-9 :+1:

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/Content.java [hunk 26]
```diff
@@ -100,4 +112,85 @@ public String toString() {
...
+   */
+  @Nonnull
+  public static AttributesMap maintainLastModified(final Asset asset,
+                                                   @Nullable final AttributesMap contentAttributes)
+  {
+    AttributesMap ca = contentAttributes;
+    if (ca == null) {
+      ca = new AttributesMap();
+    }
+    if (!ca.contains(Content.CONTENT_LAST_MODIFIED)) {
```
Comment (mcculls) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/Content.java]: Is this enough? What happens if the blob attached to the asset changes, in that case the attribute map will already contain a last-modified date and won't be updated. See https://issues.sonatype.org/browse/NEXUS-8937

File: plugins/basic/nexus-repository-nuget/src/main/java/com/sonatype/nexus/repository/nuget/internal/NugetGalleryFacetImpl.java [hunk 27]
```diff
@@ -348,55 +345,27 @@ public Payload get(final String id, final String version) {
...
 
     StorageTx tx = UnitOfWork.currentTransaction();
 
     Asset asset = findAsset(tx, id, version);
     checkState(asset != null);
 
-    final Date priorDate = asset.formatAttributes().get(P_LAST_VERIFIED_DATE, Date.class);
-    log.debug("Updating last verified date of {} {} from {} to {}", id, version, priorDate, date);
-    asset.formatAttributes().set(P_LAST_VERIFIED_DATE, date.toDate());
+    log.debug("Updating cacheInfo of {} {} to {}", id, version, cacheInfo);
```
Comment (mcculls) [inline plugins/basic/nexus-repository-nuget/src/main/java/com/sonatype/nexus/repository/nuget/internal/NugetGalleryFacetImpl.java]: Minor: only nuget appears to log this, would be nice to have the same logging in raw and maven as well

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/Content.java (hunk 26 above)
Comment (cstamas) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/Content.java]: It should be for now: there is no "update" happening here (ie. old asset loaded up as content), the contentAttributes are always created from scratch (proxy case) or always null (hosted case). NX3 always neglected uploader provided headers like last-modified, and I did not modify anything in that area.

**Comment by mcculls** [comment]
+1 thanks

File: plugins/basic/nexus-repository-nuget/src/main/java/com/sonatype/nexus/repository/nuget/internal/NugetGalleryFacetImpl.java (hunk 27 above)
Comment (cstamas) [inline plugins/basic/nexus-repository-nuget/src/main/java/com/sonatype/nexus/repository/nuget/internal/NugetGalleryFacetImpl.java]: Ack, added 5d366718f4ccc4edc0b58fe4461eaefba8e2f339

**Comment by cstamas** [comment]
Merged as d2aed3a461c786a71d40970e1b400889216e939e
